crypto/fipsmodule/kbkdf/kbkdf.c

```diff
@@ -0,0 +1,169 @@
+    uint32_t i = 0;
+    // r = sizeof(i) * BYTEBITS is listed as a parameter in SP800-108r1,
+    // specified when a counter is used as input. It's not actually used in
+    // the KDF in Feedback Mode algorithm though.
+    uint8_t zero = 0x00;
+
+    size_t n = 1 + ((L - 1) / h);  // ceil(L/h)
+
+    if (n > UINT32_MAX) {
+        // SP 800-108r1 section 4.0, n <= 2^r - 1
+        goto out;
+    }
+
+    ki = OPENSSL_malloc(ct_max(iv_len, hmac_size));
+    size_t ki_len = key_in_len;
+    if (iv != NULL && iv_len > 0) {
+        // Set k(0) to the IV.
+        memcpy(ki, iv, iv_len);
+        ki_len = iv_len;
+    }
+
+    size_t written = 0;
+    size_t to_write = out_len;
+
+    for (i = 1; i <= n; i++) {
+        if (!(i == 1 && iv_len == 0)) {
+            // If no IV was specified, we don't include k(0).
+            if (!HMAC_Update(&hmac, ki, ki_len)) {
+                goto out;
+            }
+        }
+        if (use_counter) {
+            uint32_t be_i = CRYPTO_bswap4(i);
+            if(!HMAC_Update(&hmac, (const uint8_t *)&be_i, sizeof(be_i))) {
+                goto out;
+            }
+        }
+        if (!HMAC_Update(&hmac, label, label_len)) {
+            goto out;
+        }
+        if (context != NULL && context_len > 0) {
+            if (!HMAC_Update(&hmac, &zero, sizeof(zero)) ||
+                !HMAC_Update(&hmac, context, context_len)) {
+                goto out;
+            }
+        }
+        // Another mystery in SP800-108r1; the current iteration's ki data is
+        // generated via:
+        //
+        // ki = PRF(key_in, previous ki || i || Label || 0x00 || Context || L)
+        //
+        // The test data from the CAVP Testing: SP 800-108 Key Derivation Using
+        // Pseudorandom Functions - Key-Based (KBKDF) page
+        // https://csrc.nist.gov/Projects/cryptographic-algorithm-validation-program/key-derivation
+        // seems to require us to ignore L.
+        //
+        // The spec also doesn't indicate whether L should be appended in
+        // big-endian format, or how large (in bits) it should be.
+        /*
+        if (!HMAC_Update(&hmac, (const uint8_t *)&L, sizeof(L))) {
+            goto out;
+        }
+        */
+
+        ki_len = hmac_size;
+        if (!HMAC_Final(&hmac, ki, (unsigned int *)&ki_len)) {
+            goto out;
+        }
+
+        size_t this_write = ct_min(ki_len, to_write);
+        memcpy(out_key + written, ki, this_write);
+        written += this_write;
+        to_write -= this_write;
+
+        if (to_write > 0) {
+            // Reset HMAC.
+            if (!HMAC_Init_ex(&hmac, key_in, key_in_len, digest, NULL)) {
+                goto out;
+            }
+        }
+    }
+
+    retval = 1;
+
+out:
+    FIPS_service_indicator_unlock_state();
+    if (retval) {
+        KBKDF_verify_service_indicator(digest);
+    }
+
+    if (ki != NULL) {
+        OPENSSL_cleanse(ki, ct_max(iv_len, hmac_size));
+        OPENSSL_free(ki);
+    }
```

Comment:
OPENSSL_free always cleanses memory before freeing it and handles null.
```suggestion
    OPENSSL_free(ki);
```